Quartz/Engine/Source/Utilities/Commander.cpp

```diff
@@ -0,0 +1,165 @@
+//
+// 3. Neither the name of the copyright holder nor the names of its contributors
+// may be used to endorse or promote products derived from this software without
+// specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+
+#include <Quartz/Utilities/Commander.hpp>
+#include <iostream>
+#include <utility>
+
+using namespace qz::utils;
+
+Commander::Commander(CommandBook& book, std::ostream& out, std::istream& in)
+    : m_book(book), m_out(out), m_in(in)
+{
+}
+
+Commander::~Commander() {}
+
+int CommandBook::find(const std::string& command)
+{
+	for (int j = 0; j < m_page; j++)
+	{
+		if (m_command[j] == command)
+		{
+			return j;
+		}
+	}
+	return -1;
+}
+
+void CommandBook::add(const std::string& command, const std::string& help,
+                      const std::string& permission, function f)
+{
+	int j = find(command);
+	if (j == -1)
+	{ // if command does not already exist, enter new command
+		j = m_page;
+		m_page++;
+	}
+	m_command[j]    = command;
+	m_help[j]       = help;
+	m_permission[j] = permission;
+	m_functions[j]  = std::move(f);
+	return;
+}
+
+int CommandBook::getPage() { return m_page; }
+
+int Commander::help(
+    const std::array<std::string, qz::utils::MAX_ARGUMENTS_NUMBER>&& args)
+{
+	if (args[0] == "")
+	{
+		m_out << "Type /help [command] to learn more about a command \nType "
+		         "/list for a list of available commands\n";
+		return 1;
+	}
+	else if (args[0] == "help")
+	{
+		m_out << "Type /help [command] to learn more about a command \n";
+		return 1;
+	}
+	else if (args[0] == "list")
+	{
+		m_out << "Lists available commands\n";
+		return 1;
+	}
+	int j = m_book.find(args[0]);
+	if (j == -1)
+	{
+		m_out << "Command \"" + args[0] + "\" not found \n";
+		return -1;
+	}
+	else
+	{
+		m_out << m_book.m_help[j];
+		return 1;
+	}
+}
+
+int Commander::run(const std::string& command,
+                   const std::array<std::string, MAX_ARGUMENTS_NUMBER>&& args)
+{
+	// Check for built in functions
+	if (command == "help")
+	{
+		return this->help(std::move(args));
+	}
+	else if (command == "list")
+	{
+		this->list();
+		return 1;
+	}
+	// If no built in functions match, search library
+	int j = m_book.find(command);
```

Comment:
Again, is there a better name we could use for `j` here?

Also the variable can be `const`
```suggestion
	const int j = m_book.find(command);
```

Reply:
See above